internal/app/backend/synchronizer_client.go

```diff
@@ -58,15 +59,35 @@ func (sc *synchronizerClient) evaluate(ctx context.Context, id string, proposals
 		return nil, err
 	}
 
-	resp, err := sc.synchronizer.EvaluateProposals(ctx, &ipb.EvaluateProposalsRequest{
-		Id:      id,
-		Matches: proposals})
-	if err != nil {
-		return nil, err
+	if len(proposals) == 0 {
+		return proposals, nil
+	}
+
+	stream, err := sc.synchronizer.EvaluateProposals(ctx)
+	for _, proposal := range proposals {
+		if err = stream.Send(&ipb.EvaluateProposalsRequest{Id: id, Match: proposal}); err != nil {
+			return nil, err
+		}
+	}
+	if err = stream.CloseSend(); err != nil {
+		logger.Errorf("failed to close the send stream: %s", err.Error())
+	}
+
+	results := []*pb.Match{}
```

Comment:
I see you following a separate pattern on the Evaluate interface - of sending over a separate goroutine and receiver blocking - any reason for these different patterns? I guess in both places, we dont expect partial responses so serializing the stream send / receive better?

Reply:
Edited